### PR TITLE
feat: visual regression testing and dev server app

### PR DESCRIPTION
**Pull request by kellyjosephprice**
[Demo][demo] | 🎫 [Ticket][ticket]
:---:|:---:

### 🧰  Changes

Adds an example app to develop against. To start the dev server:

```
npm run start
```

And you can also run visual regression tests against it:

```
npm run test.browser
```

This runs `jest-image-snapshot` against the dev server. It uses saved
screenshots in `__tests__/browser/__image_snapshots__` to compare. If it
finds differences, it'll save diffs to
`__tests__/browser/__image_snapshots__/__diff_output__`.

Currently, there's little guarantee that your environment will match the
Github Actions environment. So, if you need to update snapshots, the
steps are:

* `sed -i -e 's/test.browser/test.browser -- -u/' .github/workflows/ci.yml`
* commit and push to github
* download 'Artifacts' from last build
* commit updated snapshots from 'Artifacts'

---

### TODO

- [x] update snapshots from CI
- [x] trim CI adornments

---

### Next steps

- [x] Add PR apps via heroku pipeline. Adding PR apps will make it nice
  to snapshot changes in PRs.
- [ ] dockerize puppeteer. Dockerizing the puppeteer browser should
  reduce broswer rendering fragility with the tests, and make it easier
  to create snapshots that match CI.

---

### Things that are broken

- [ ] testing snapshots currently fails locally. Presumably until we use
  docker locally, we'll be out of sync with CI.
- [ ] `npm run test.browser` seems to have broken jest's cache. You need
  to run `npx jest --clearCache` between runs
- [ ] `npm run test.browser.watch` does not kill the dev server. We
  typically won't be running it in watch mode.

[demo]: https://markdown-pr-94.herokuapp.com/#
[ticket]: https://app.asana.com/0/1199089258199423/1199546355080189/f

### CODE REVIEW

**Comment by kellyjosephprice** [comment]
@rafegoldberg want to try out `npm start`? I think it's doing what it's more or less supposed to. You can:

* load docs from https://rdmd.readme.io/docs/code-blocks (I just did an export from the site, it'd be nice to sync them)
* edit docs live

Next step is writing some snapshot tests.

Also, thanks to this year, this bro has a different meaning to me:
image

**Comment by rafegoldberg** [comment]
😹 lol if you like him, you'll love his dad. (more pertinently: excited to check this out!)

File: webpack.dev.config.js
Comment (rafegoldberg) [inline webpack.dev.config.js:2]:
```suggestion
/* eslint-disable import/no-extraneous-dependencies
 */
const path = require('path');
```

File: webpack.dev.config.js
Comment (rafegoldberg) [inline webpack.dev.config.js:28]: What's this for?

File: webpack.config.js
Comment (rafegoldberg) [inline webpack.config.js:92]: I think this is a little confusing. [The conventional setup](https://webpack.js.org/guides/production/#setup) is to split the common config out in to a _third_ file, instead of munging the main config's exports!
```
@readme/markdown
 ├ webpack.config.js
 ├ webpack.common.js
 └ webpack.dev.js
```

File: package.json
Comment (rafegoldberg) [inline package.json:81]: What do we need Lodash for?

**Review by rafegoldberg — CHANGES_REQUESTED** [review]
I left a couple of questions and smaller suggestions below, but overall I'm psyched for this! My only real ask is that we ignore the `example/img/emojis/` directory in git.

File: webpack.dev.config.js
Comment (kellyjosephprice) [reply webpack.dev.config.js:28]: Whoops. I was trying to figure out why webpack was so slow. I had added sourcemaps.

File: package.json
Comment (kellyjosephprice) [reply package.json:81]: https://github.com/readmeio/markdown/pull/94/files#diff-58d8608f00398d6dfc613a4961f1641a01f9320dc5e72a19f5689d2c9a6e4cbeR27

File: example/Fixtures/Syntax/index.js
Comment (rafegoldberg) [inline example/Fixtures/Syntax/index.js:27]: Any reason to not use vanilla JS here? Then we can nix Lodash!
```suggestion
  const name = sym;
```

File: package.json
Comment (rafegoldberg) [reply package.json:81]: [Any reason to not use vanilla JS here?](https://github.com/readmeio/markdown/pull/94#discussion_r543694504)

File: example/Fixtures/Syntax/index.js
Comment (kellyjosephprice) [reply example/Fixtures/Syntax/index.js:27]: No, I can write my own lowerCase.

Comment (rafegoldberg) [reply example/Fixtures/Syntax/index.js:27]: Just use the [built-in **`string.toLowerCase()`**](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLowerCase)! (Sorry, I forgot to paste it in to my suggestion!)

Comment (kellyjosephprice) [reply example/Fixtures/Syntax/index.js:27]: I want `lowerCase` => `lower case`.

Comment (rafegoldberg) [reply example/Fixtures/Syntax/index.js:27]: Ohhhhhh shit, guess I shoulda read the lodash docs before I commented....

Comment (rafegoldberg) [reply example/Fixtures/Syntax/index.js:27]: In that case, feel free to leave the dependency.

File: Makefile
Comment (kellyjosephprice) [reply Makefile:4]: IIUC, I thought `.PHONY` was to protect against the case were you create a file the same name as an unrelated make target. I imagine it's unlikely we'd ever add an `./emoji` file :shrug:

File: webpack.dev.config.js
Comment (kellyjosephprice) [reply webpack.dev.config.js:28]: Incidentally, loading the markdown files with sourcemaps enabled causes this build to go from around <500ms to 10-15 seconds. Very annyoing.

File: package.json
Comment (kellyjosephprice) [inline package.json:19]: @rafegoldberg how does this make you feel?

The way it's setup you should be able to run `npx jest path/to/any/file` and it works. The downside is that the package script has to know the paths. 

`jest` has the concept of `packages`, but they're path based. So you could do something like:

```
"test.browser": "jest --projects __test__/browser",
```

and put a `jest.config.js` in `__test__/browser`. But, the config gets annoying because `jest`'s `<rootDir>` gets set to that dir. It just seems easier to keep them more similar.

File: Makefile
Comment (rafegoldberg) [reply Makefile]: Ahhh, okay. My understanding of makefiles is cursory (at best) so educate me cap'n!

File: package.json
Comment (rafegoldberg) [reply package.json:19]: Thinking about this more today; let's chat "offline" today about the scripts/file structure/naming conventions in this PR?

File: .github/workflows/ci.yml
Comment (erunion) [inline .github/workflows/ci.yml:22]: What's this for?

File: __tests__/browser/markdown.test.js
Comment (erunion) [inline __tests__/browser/markdown.test.js:27]: Might want to use `it.each` instead. I've had Jest be really spotty in the past if you run tests within a loop like this.

File: jest.common.js
Comment (erunion) [inline jest.common.js:3]: Can probably slim this down since you're just dealing with `scss` here.

File: package.json
Comment (erunion) [inline package.json:76]: What're you using core-js for?

File: .github/workflows/ci.yml
Comment (kellyjosephprice) [reply .github/workflows/ci.yml:22]: Providing unicode (non @readme emojis) support.

Ideally, we'd run the browser test in a docker container, so we can duplicate the env that ci uses.

File: __tests__/browser/markdown.test.js
Comment (kellyjosephprice) [reply __tests__/browser/markdown.test.js:27]: TIL.

File: package.json
Comment (kellyjosephprice) [reply package.json:76]: Trying to get `async`/`await` support? I was opting not to use `babel-polyfill` since it's deprecated. In that case, I should probably replace `babel-polyfill` elsewhere in the repo?

File: Makefile
Comment (rafegoldberg) [inline Makefile:6]: Does this need to be split in to two tasks?

```suggestion
emojis: ## Install our emojis
	node_modules/@readme/emojis
```

File: package.json
Comment (rafegoldberg) [inline package.json:22]: Should we prefix this with the Jest cache busting command you showed me? (I recently found out you can add a `pre` and `post` tasks for _any_ script and NPM will automatically execute 'em in order when you run the main task!)

```suggestion
    "pretest.browser": "jest --clearCache",
    "test.browser": "jest --runInBand --testPathPattern '__tests__/browser'",
```

File: .gitignore
Comment (rafegoldberg) [inline .gitignore:12]: I think we should ignore the entire snapshots directory:

```suggestion
__image_snapshots__
```

Otherwise local tests get diffed against the GitHub action's output (or users could potentially push snapshots updated locally that might break GitHub's comparison!) which gives a pretty silly failing diff:

image

But ignoring this folder would actually break diff comparison on the remote, since it would _never_ have anything to compare against, right? So maybe we can get around this by storing the locally generated snapshots in a separate directory, and then just ignore _that?_ That way, GitHub would always diff against remote snapshots!

**Review by rafegoldberg — CHANGES_REQUESTED** [review]
I don't think we need Docker to get these diffs to work...

File: package.json
Comment (rafegoldberg) [inline package.json:23]: We can probably just get rid of this script in favor of executing `npm run test.browser -- --watch` directly!
```suggestion
```

File: Makefile
Comment (kellyjosephprice) [reply Makefile:6]: Splitting it in two was so:

* we can type `make emojis`
* it only runs if `example/img/emojis` is older than `node_modules/@readme/emojis`

But then again, it probably doesn't matter.

File: package.json
Comment (kellyjosephprice) [reply package.json:22]: oooh TIL. very cool.

File: Makefile
Comment (rafegoldberg) [reply Makefile:6]: lol no yours is better; like I said, I know not about makefiles!

File: .gitignore
Comment (kellyjosephprice) [reply .gitignore:12]: So wrapping `test.browser` in a docker container would let us run the specs in the same environment. So, we should be able to output the same snapshots as on CI.

But yea, we could use a different snapshot dir for CI?

Comment (kellyjosephprice) [reply .gitignore:12]: > So wrapping test.browser in a docker container would let us run the specs in the same environment. So, we should be able to output the same snapshots as on CI.

And between dev's laptops

Comment (rafegoldberg) [reply .gitignore:12]: Again, that sounds much better in the long run, but I'm not sure how long it would take to dockerize this stuff? Maybe we can do that work in a separate PR, once we've got this stuff merged in and running for a little bit? Your call though.

File: Makefile
Comment (kellyjosephprice) [reply Makefile:6]: Heh, I think the original use case was like incrementally compiling c?

```
program.exe: program.o library.o
  gcc --some command

program.o: program.c
  gcc --some other command

library.o: library.c
  gcc --some other other command
```

If you made some change to `program.c` and ran `make` it would skip recompiling library.c?

Comment (rafegoldberg) [reply Makefile:6]: Ahh, I gotchya. I don't think that kind of smart comparison would apply in our case though, since we manually running that make task as a one-off? But like you said: "it ~probably~_definitely_ doesn't matter".

File: .gitignore
Comment (kellyjosephprice) [reply .gitignore:12]: I'm tempted to say a couple hours. But the reality is probably a couple days?

**Comment by rafegoldberg** [comment]
## This PR was released!

🚀 Changes included in [<b>v6.25.0-next.1</b>](/readmeio/markdown/releases/tag/v6.25.0-next.1)
<br>

**Comment by rafegoldberg** [comment]
These changes are super exciting. I should note that we still need to fix this workflow up so it works seamlessly between local dev environments and remote CI runs.

**Comment by rafegoldberg** [comment]
## This PR was released!

🚀 Changes included in [<b>v6.25.0</b>](/readmeio/markdown/releases/tag/v6.25.0)
<br>

**Pull request merged**
